proposals/function-delegate.md

```diff
@@ -0,0 +1,195 @@
+# Function delegate
+
+* **Type**: Design proposal
+* **Author**: Marcin Moskała
+* **Contributors**: 
+* **Status**: Proposition
+* **Prototype**: Proposition
+
+## Feedback 
+
+Discussion of this proposal is held in [this issue](https://github.com/Kotlin/KEEP/issues/25).
+
+## Summary
+
+Support function delegates - delegae that can be used on to delegate function calls.
+
+## Description
+
+Allow to use function type as a function delegate:
+
+```
+interface MainView {
+    fun clearList()
+}
+
+class ListAdapter {
+    fun clear() {}
+}
+
+class MainActivity : MainView {
+    val adapter = ListAdapter()
+    
+    override fun clearList() by adapter::clear
+}
+```
+
+## Motivation / use cases
+
+Two main motivations behind this feature are:
+ * Simplify function calls passing that is opened for function signature changes. 
+ * Allow wrapper functions with state to add catching, synchronization and other functionalities that needs state. 
+Let's discusse them one after another. 
+
+### Function calls passing 
+
+When we are keeping strict architectures, like MVP or Clean Architecture, we often need to pass call from one class to another. For example, when Presenter is deciding that list needs to be changed it needs to call Activity function which is passing this call to adapter which is managing elements on list:
+
+```
+interface MainView {
+    fun deleteListElementAt(position: Int)
+}
+
+class ListAdapter {
+    fun deleteAt(position: Int) {}
+}
+
+class MainActivity : MainView {
+    val adapter = ListAdapter()
+    
+    override fun deleteListElementAt(position: Int) {
```

Comment:
It changes nothing, and I am not using single expression functions for expressions that are returning `Unit` this is misleading for a person who is reading it.

Reply:
>It changes nothing

Current example takes 3 lines of code without delegate, with delegate it takes 1. But in simple cases delegate only omits function arguments in compare to expression body `=`.

>and I am not using single expression functions for expressions that are returning `Unit` because this is misleading for a person who is reading it.

- In your own examples delegate hides return type `Unit` as well: `override fun deleteListElementAt(position: Int) by adapter::deleteAt`
- You can always specify return type explicitly if you want `override fun deleteListElementAt(position: Int): Unit = adapter.deleteAt(position)`

Just want to remove bias from the syntax comparison, so far I'm positive about this proposal :)

Reply:
Sound fair. I will change it then